File: framework/test2/common/intbreak/intbreak.py
#!/usr/bin/env python3
from collections import Counter, defaultdict, deque
import heapq
import json
import logging
import math
import random
import time
from typing import *
import sys

from soda.leetcode.bitree import *
from soda.leetcode.graph import *
from soda.leetcode.linklist import *
from soda.unittest.util import init_logging

logger = logging.getLogger(__name__)

# step [1]: implement class Solution
# class Solution: pass
class Solution:
    def integerBreak(self, n: int) -> int:
        self.memo = [0] * 59
        return self.solve(n)

    def solve(self, n):
        if n == 1:
            return 1
        if self.memo[n] > 0:
            return self.memo[n]
        res = 0
        for i in range(1, n):
            res = max(i * (n-i), i * self.solve(n-i), res)
        self.memo[n] = res
        return res

if __name__ == '__main__':
    init_logging()
    from soda.unittest import Validators
    from soda.unittest.work import TestWork

    # step [2]: setup function
    # Attention! FUNCTION must use type hint, including arguments and return type
    work = TestWork(Solution().integerBreak)
    # OR use struct tester
    # work = TestWork.forStruct(CLASS)

    # step [3]: setup other options
    # work.validator = (e,r) => bool
    work.compareSerial = True
    work.run()

File: framework/test2/common/intbreak/intbreak.php
<?php
require_once 'Soda/DS/index.php';
require_once 'Soda/Unittest/index.php';

use Soda\DS;
use Soda\Unittest\TestWork;
use Soda\Unittest\Utils;

class Solution {

    private $memo;

    /**
     * @param Integer n
     * @return Integer
     */
    function integerBreak($n) {
        $this->memo = array_fill(0, 59, 0);
        return $this->solve($n);
    }

    private function solve($n) {
        if ($n == 1) {
            return 1;
        }
        if ($this->memo[$n] > 0) {
            return $this->memo[$n];
        }
        $res = 0;
        for ($i = 1; $i < $n; ++$i) {
            $res = max($i * ($n-$i), $i * $this->solve($n-$i), $res);
        }
        $this->memo[$n] = $res;
        return $res;
    }
}

$work = TestWork::create(new Solution(), 'integerBreak', $argv[0]);
// $work->validator = fn($e, $r) => <boolean result>
$work->compareSerial = true;
echo $work->run(Utils::fromStdin());
